ElectricalCircuitSimulator/SimulationEngine/src/Matrix.cpp

```diff
@@ -49,167 +65,149 @@ namespace SimulationEngine {
+                cout << "\t" << m_pMatrix[iRow][iColumn];
             }
+            cout << "]\n";
         }
-
-        return pNewMatrix;
+        cout << endl;
     }
 
-    void Matrix::printMatrix() const {
-        int iRow;
-        int iColumn;
+    void Matrix::linearSystemSolver(const Matrix& oSourceVector, const PLU_Factorization& oPLU, Matrix& oVoltageVector) {
+        int iI1;
+        int iI2;
+        int iIndex;
+        int iNumRows = oSourceVector.m_iNumRows;
+        Matrix oX(iNumRows, 1); // UX = Y
+        Matrix oY(iNumRows, 1); // LY = B_Permuted
+        Matrix oB_Permuted(iNumRows, 1);
 
-        for (iRow = 0; iRow < m_iNumRows; iRow++) {
-            cout << "[" << m_oMatrix[iRow][0];
-            for (iColumn = 1; iColumn < m_iNumColumns; iColumn++) {
-                cout << "\t" << m_oMatrix[iRow][iColumn];
+        // Apply row permutations to B to get B_Permuted
+        for (iI1 = 0; iI1 < iNumRows; ++iI1) {
+            iIndex = (int)round(oPLU.getP()(iI1, 0));
+            oB_Permuted(iI1, 0) = oSourceVector(iIndex, 0);
+        }
+
+        // Forward substitution to solve LY = B_Permuted
+        for (iI1 = 0; iI1 < iNumRows; ++iI1) {
+            oY(iI1, 0) = oB_Permuted(iI1, 0);
+            for (iI2 = 0; iI2 < iI1; ++iI2) {
+                oY(iI1, 0) -= oPLU.getL()(iI1, iI2) * oY(iI2, 0);
             }
-            cout << "]\n";
         }
-        cout << endl;
+
+        // Backward substitution to solve UX = Y
+        for (iI1 = iNumRows - 1; iI1 >= 0; --iI1) {
+            oX(iI1, 0) = oY(iI1, 0);
+            for (iI2 = iI1 + 1; iI2 < iNumRows; ++iI2) {
+                oX(iI1, 0) -= oPLU.getU()(iI1, iI2) * oX(iI2, 0);
+            }
+            if (fabs(oPLU.getU()(iI1, iI1)) > ZERO) {
+                oX(iI1, 0) /= oPLU.getU()(iI1, iI1);
+            } else {
+                oX(iI1, 0) = 0; // If a diagonal on the U matrix is 0, it's due to the ground node being included in the matrix, and is effectively infinity, resulting in oX = 0 for this row.
+            }
+        }
+
+        // Apply column permutations to X using Q to get X_Final
+        for (iI1 = 0; iI1 < iNumRows; ++iI1) {
+            iIndex = (int)round(oPLU.getQ()(iI1, 0));
+            oVoltageVector(iIndex, 0) = oX(iI1, 0);
+        }
+
+#ifdef MATRIX_PRINT
+        cout << "X Vector:" << endl;
+        oX.printMatrix();
+        cout << "Y Vector:" << endl;
+        oY.printMatrix();
+        cout << "B Permuted Vector:" << endl;
+        oB_Permuted.printMatrix();
+#endif
     }
 
     // Function to perform PLU factorization with row and column pivoting
     // Factors a matrix into P*A*Q = L*U, where P = Row Permutation Matrix, and Q = Column Permutation Matrix
-    PLU_Factorization* Matrix::runPLU_Factorization() const {
-        PLU_Factorization* pResults = new PLU_Factorization();
+    void PLU_Factorization::runPLU_Factorization(const Matrix& oConductanceMatrix) {
```

Comment:
I'd probably rename the variable. it doesn't *have* to be a conductance matrix. This is just a PLU_Factorization function. Any matrix that can be decomposed will do.